src/RequestInformation.cs

```diff
@@ -157,29 +156,52 @@ public Uri URI
         public void AddQueryParameters<T>(T source)
 #endif
         {
-            if (source == null) return;
-            foreach (var property in typeof(T)
-                                        .GetProperties()
-                                        .Select(
-                                            x => (
-                                                Name: x.GetCustomAttributes(false)
-                                                    .OfType<QueryParameterAttribute>()
-                                                    .FirstOrDefault()?.TemplateName ?? x.Name,
-                                                Value: x.GetValue(source)
-                                            )
-                                        )
-                                        .Where(x => x.Value != null &&
-                                                    !QueryParameters.ContainsKey(x.Name!) &&
-                                                    (x.Value is not ICollection collection || collection.Count > 0))) // no need to add empty collection
+            if(source == null) return;
+
+            foreach(var propertyInfo in typeof(T).GetProperties())
             {
-                QueryParameters.AddOrReplace(property.Name!, property.Value!);
+                var attributes = propertyInfo.GetCustomAttributes(false);
+
+                QueryParameterAttribute? queryParameterAttribute = null;
+                foreach(var attribute in attributes)
+                {
+                    if(attribute is QueryParameterAttribute qpAttribute)
+                    {
+                        queryParameterAttribute = qpAttribute;
+                        break;
+                    }
+                }
+
+                if(queryParameterAttribute is not null)
+                {
+                    var templateName = queryParameterAttribute.TemplateName ?? propertyInfo.Name;
```

Comment:
This may not be equivalent to the previous code, in the scenario where the `queryParameterAttribute` isn't present, we should still fall back to the `propertyInfo.Name` which won't happen as we won't enter the if clause(check currently failing tests).